apache2/msc_pcre.c

```diff
@@ -21,11 +21,15 @@
 static apr_status_t msc_pcre_cleanup(msc_regex_t *regex) {
     if (regex != NULL) {
         if (regex->pe != NULL) {
+            if (((pcre_extra *)regex->pe)->flags & PCRE_EXTRA_EXECUTABLE_JIT) {
```

Comment:
This's the code from ModSecurity v3. Is it correct? If yes, maybe it's better to just borrow it from there? It seems more straightforward.

https://github.com/SpiderLabs/ModSecurity/blob/357c14000350234beb017ea08f01dc7fed27ed4a/src/utils/regex.cc#L41-L66

Reply:
This is indeed more straightforward and effective (branching will happen at compile time instead of run time).
I did not do it this way initially because it (at least theoretically) can be so that `WITH_PCRE_STUDY` and `WITH_PCRE_JIT` **are** defined, and `pcre_study` **was** called, but returned `NULL` (man page for `pcre_study` says `If the function returns NULL, either it could not find any additional information, or there was an error. You can tell the difference by looking at the error value. It is NULL in first case.`). In this case `msc_pregcomp_ex` will allocate `pcre_extra` with `malloc`/`pcre_malloc`, it does so because it uses `pcre_extra` to set matching limits.
The relevant code is here

https://github.com/SpiderLabs/ModSecurity/blob/aa18ec7/apache2/msc_pcre.c#L68-L87

Reply:
But are you aware that you basically replicated the same (simple) check that's inside `pcre_free_study` itself? A function call to libpcre library is going to happen either way (`pcre_free_study` or `pcre_free`), so I don't think there's any performance gain.

https://vcs.pcre.org/pcre/code/trunk/pcre_study.c?revision=1639&view=markup

(extra ifdefs omitted for clarity)
```c
void pcre_free_study(pcre_extra *extra)
{
if (extra == NULL)
  return;
#ifdef SUPPORT_JIT
if ((extra->flags & PCRE_EXTRA_EXECUTABLE_JIT) != 0 &&
     extra->executable_jit != NULL)
  PRIV(jit_free)(extra->executable_jit);
#endif
PUB
```

Reply:
> But are you aware that you basically replicated the same (simple) check that's inside `pcre_free_study` itself

Yes, I've even written it in commit message and issue description myself:

> and in fact `pcre_free_study` works the same way: it checks that flag, if it is not set, it calls `pcre_free` on `pcre_extra` pointer 

> so I don't think there's any performance gain

I worry mostly about safety, not performance here (including trying to avoid calling deallocating function from one allocator on pointer produced by other allocator. PCRE allows to replace `pcre_malloc` and `pcre_free` with custom allocation functions, which is used, [for example, by nginx](https://github.com/nginx/nginx/blob/4bf4650f2f10f7bbacfe7a33da744f18951d416d/src/core/ngx_regex.c#L74)).

And I don't know how correct it is to rely on current implementation of `pcre_free_study`, to expect it to stay the same. 

Even safer option here might be also to not rely on `PCRE_EXTRA_EXECUTABLE_JIT` flag, but add a new field to `msc_regex_t`.
What do you think @zimmerle ?

Reply:
There're three cases:

1) The code is not using JIT, `VERSION_NGINX` is not defined. The `pcre_extra` struct is allocated with plain `malloc`, and has to be freed with plain `free` (technically, you can mix and match allocator functions if you don't touch the global variables `pcre_malloc` and `pcre_free`, but that obviously smells and is fragile).
2) The code is not using JIT, `VERSION_NGINX` is defined. Memory is allocated with `pcre_malloc`, and both `pcre_free` and `pcre_free_study` can be used.
3) The code is using JIT. In this case, the memory _has_ to be freed by `pcre_free_study`. Using `pcre_free` would cause memory leak, as JIT-compiled code woudn't be freed. 

The complexity comes from distinguishing cases 1 and 3. If we simply used `pcre_malloc` all the time, we could've just used `pcre_free_study`. 

The `ifdef` was introduced in 154e80750. I think that `ifdef` complicates things unnecessarily, and we can drop it and simply use `pcre_malloc`/`pcre_free`. After all, these two variables are just function pointers initialized to `malloc` and `free` correspondingly.

Reply:
I researched it a bit more and found this doc https://www.pcre.org/original/doc/html/pcreapi.html#SEC13 (which turned out to be a web version of `man pcreapi`). After reading it and after more discussion with @WGH- I got more convinced that it is worth changing the code a bit more to make the fix a bit cleaner and make some refactoring.

The doc says that it is recommended to replace usages of `pcre_free()` to deallocate `pcre_extra` with `pcre_free_study()`:
> When you are finished with a pattern, you can free the memory used for the study data by calling pcre_free_study(). This function was added to the API for release 8.20. For earlier versions, the memory could be freed with pcre_free(), just like the pattern itself. This will still work in cases where JIT optimization is not used, but it is advisable to change to the new function when convenient.

So I agree with @WGH-  that using regular `malloc` and `free` for `pcre_extra` seem to be of no use, and that it is correct and more clean to use `pcre_malloc` instead, and in this case `pcre_free_study` can be used to deallocate it unconditionally.

The doc also mentions there is a flag `PCRE_STUDY_EXTRA_NEEDED` that can be passed to `pcre_study()` to ask it to allocate `pcre_extra` anyway, even if studying `did not find any additional information` - which may be useful for us since `pcre_extra` is needed anyway to set match limits on it.

I created and alternative version of this branch with described approach,in fact only the second commit is changed, it can be found here https://github.com/seclab-msu/ModSecurity/commit/5859417502d3058599c0c84261db5a8fe886c7c1
That branch passes tests (run by `tests/run-unit-tests.pl`) as well as original branch of this PR, I also tested running modsec as Apache module and also built it as Nginx module and tested that, it seemed to work OK.
I am ready to replace branch of this pull request with that one if it's better @zimmerle

Reply:
I replaced the branch of this PR with aforementioned alternative version of the fix